### PR TITLE
OAM-218: Many program ids for GET API for orderables.

### PR DESCRIPTION
**Pull request by tsznaj**

### CODE REVIEW

**Comment by tsznaj** [comment]
Integration tests have still to be fixed.

File: src/main/java/org/openlmis/referencedata/web/OrderableSearchParams.java [hunk 1]
```diff
@@ -75,4 +73,9 @@ public Set<UUID> getTradeItemId() {
   public boolean getIncludeQuarantined() {
     return includeQuarantined;
   }
+
+  @Override
+  public Set<String> getProgramCodes() {
+    return Collections.singleton(programCode);
```
Comment (sradziszewski) [inline src/main/java/org/openlmis/referencedata/web/OrderableSearchParams.java:79]: should it always be like that (returning singleton?)

Comment (tsznaj) [reply src/main/java/org/openlmis/referencedata/web/OrderableSearchParams.java:79]: I think so, as we are not supposed to change POST search endpoint, where we can get still only one program code.
Or maybe we should change it to have the ability to pass many program codes to POST search endpoint as well?

File: src/main/java/org/openlmis/referencedata/repository/custom/impl/OrderableRepositoryImpl.java [hunk 2]
```diff
@@ -229,16 +232,20 @@ <E> TypedQuery<E> prepareQuery(SearchParams searchParams, CriteriaQuery<E> query
...
+      Set<String> programCodes = searchParams.getProgramCodes() != null
+          ? getProgramCodesLowerCase(searchParams)
+          : Collections.emptySet();
+      if (!programCodes.isEmpty()) {
         Join<Orderable, ProgramOrderable> poJoin = root.join(PROGRAM_ORDERABLES, JoinType.INNER);
         Join<ProgramOrderable, Program> programJoin = poJoin.join(PROGRAM, JoinType.INNER);
-        where = builder.and(where, builder.equal(builder.lower(programJoin.get(CODE).get(CODE)),
-            searchParams.getProgramCode().toLowerCase()));
+        where = builder.and(where, builder.lower(programJoin.get(CODE).get(CODE))
+            .in(getProgramCodesLowerCase(searchParams)));
```
Comment (pwargulak) [inline src/main/java/org/openlmis/referencedata/repository/custom/impl/OrderableRepositoryImpl.java:248]: Use `programCodes` from few lines above.

File: src/main/java/org/openlmis/referencedata/repository/custom/impl/OrderableRepositoryImpl.java [hunk 3]
```diff
@@ -299,10 +314,11 @@ private Query getLastUpdatedQuery(SearchParams searchParams, boolean count) {
     String queryCondition;
 
     if (null != searchParams) {
-      if (null != searchParams.getProgramCode()) {
+      if (null != searchParams.getProgramCodes()) {
```
Comment (pwargulak) [inline src/main/java/org/openlmis/referencedata/repository/custom/impl/OrderableRepositoryImpl.java:317]: Very similar pattern is in `org.openlmis.referencedata.repository.custom.impl.OrderableRepositoryImpl#prepareParams` too - make them consistent. 
Consider moving `null != searchParams.getProgramCodes()` to the `getProgramCodesLowerCase` method.

File: src/main/java/org/openlmis/referencedata/repository/custom/impl/OrderableRepositoryImpl.java [hunk 4]
```diff
@@ -330,6 +346,19 @@ private Query getLastUpdatedQuery(SearchParams searchParams, boolean count) {
...
 
+  private String generateProgramCodesText(Set<String> programCodesLowerCase) {
+    StringJoiner joiner = new StringJoiner(", ");
+    for (String programCode : programCodesLowerCase) {
+      StringBuilder builder = new StringBuilder();
+      builder
+          .append('\'')
+          .append(programCode)
+          .append('\'');
+      joiner.add(builder.toString());
```
Comment (pwargulak) [inline src/main/java/org/openlmis/referencedata/repository/custom/impl/OrderableRepositoryImpl.java:357]: Not a request to change, but I'm almost sure the more convenient syntax: `"'" + programCode + "'"` is going to be optimzied to StringBuilder (or better).

File: src/main/java/org/openlmis/referencedata/web/OrderableSearchParams.java (hunk 1 above)
Comment (pwargulak) [reply src/main/java/org/openlmis/referencedata/web/OrderableSearchParams.java:80]: So `OrderableSearchParams` supports search by single programCode, but it implements `OrderableRepositoryCustom.SearchParam` which supports multiple `programCodes`. The `getProgramCodes` is part of `OrderableRepositoryCustom.SearchParam` contract, and is JsonIgnore becacuse there is no `programCodes` property, no setter to call when serializing/deserializing.

Hmm, makes sense.

**Review by pwargulak — CHANGES_REQUESTED** [review]
Left some questions

**Review by sradziszewski — APPROVED** [review]
nothing from me except Piotr's comments

**Comment by tsznaj** [comment]
Piotr cannot technically approve now here - but he has approved.

**Pull request merged**